Add ability to group related service configurations

warappa opened the pull request:
Add `IServiceCollectionModule` and `ServiceCollectionModuleExtensions`.

Fixes #37867

warappa committed:
Adds ability to group related service configurations

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=37868) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=37868) before we can accept your contribution.<br/><br/>:x: warappa [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=37868)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=37868) it.</sub>

warappa committed:
Return void in module's ConfigureServices method

Applied feedback of @KalleOlaviNiemitalo.
The previous API contract suggested an option that should not exist nor be utilized: replacing the provided IServiceCollection instance with a new one. Also nearer by Startup.cs' ConfigureServices (standard) method signature.

maryamariyan commented:
Thanks for your contribution, @warappa.

This change requires API review and approval. Please see the process [here](https://github.com/dotnet/runtime/blob/master/docs/project/api-review-process.md).

maryamariyan commented:
closing for now, and you could reopen once the API is approved.
